awx/sso/fields.py

```diff
@@ -74,6 +76,71 @@ def get_depends_on(self):
         return res
 
 
+class _Forbidden(Field):
+    default_error_messages = {
+        'invalid': _('Invalid field.'),
+    }
+
+    def run_validation(self, value):
+        self.fail('invalid')
+
+
+class HybridDictField(fields.DictField):
+    """A DictField, but with defined fixed Fields for certain keys.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.allow_blank = kwargs.pop('allow_blank', False)
+
+        fields = [
+            (field_name, obj)
+            for field_name, obj in self.__class__.__dict__.items()
+            if isinstance(obj, Field) and field_name != 'child'
+        ]
+        fields.sort(key=lambda x: x[1]._creation_counter)
+        self._declared_fields = collections.OrderedDict(fields)
+
+        super().__init__(*args, **kwargs)
+
+    def to_representation(self, value):
+        fields = copy.deepcopy(self._declared_fields)
```

Comment:
We've had some major problems with `deepcopy()` in the past, such that any time it's used the usage sticks out to me. It's probably fine?  Are there deep references here that need the deep copy?

Reply:
This stuff is borrowed directly from DRF's implementation of `Serializer`, folded together a bit since we don't need the full generality.  We could probably get away with not deepcopy-ing, since it seems unlikely that we'll dynamically add fields to these things, per https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L380-L383